server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/PendingListenersQueue.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.NotMasterException;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class PendingListenersQueue {
+
+    private static final Logger logger = LogManager.getLogger(PendingListenersQueue.class);
+
+    private record PendingListener(long index, ActionListener<Void> listener) {}
+
+    private final ThreadPool threadPool;
+    private final Queue<PendingListener> pendingListeners = new LinkedList<>();
+    private volatile long completedIndex = -1;
+
+    public PendingListenersQueue(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    public void add(long index, ActionListener<Void> listener) {
+        synchronized (pendingListeners) {
+            pendingListeners.add(new PendingListener(index, listener));
+        }
+    }
+
+    public void complete(long index) {
+        advance(index);
+        executeListeners(completedIndex, true);
+    }
+
+    public void completeAllAsNotMaster() {
+        completedIndex = -1;
```

Comment:
This looks unsafe to me, as in if `advance` and `completeAllAsNotMaster` runs on different threads, we risk advance setting `completedIndex` to an index after we set it to -1 here?

I am not exactly sure why we reset the `indexGenerator` in `DesiredBalanceShardsAllocator`, could it not continue where it left in case the node becomes master again? That would avoid the reset here, simplifying I think.

Reply:
Tracked in #91386. @idegtiarenko could you take a look at this?

Reply:
It is important co complete the listeners so that we do not have stuck requests if the node is no longer master.
I also think it is worth resetting the desired balance to empty/initial case as there could be a lot of changes to the routing table by the time the node is elected as a master again.I guess it is fine not to reset the index here (the one in desired balance allocator should not be resetted as well)